### ISSUE TITLE
smtp app credentials in plain text or gmail ip whitelisting

### ISSUE THREAD

**Issue by rojomisin** [GoogleCloudPlatform/jit-access#50]
It seems like the smtp credentials would be stored plain text, which is not ideal.  Is there an easy way to:
- have java use an encrypted version of the password
or
- have app.yaml accept a secrets manager path
or
- if using IP whitelisting only, not smtp authentication, can the JIT access page display the current CIDR block of the app engine app, to use in the gmail smtp relay settings?  Otherwise we would have to input potentially [60-70 CIDR ranges](https://www.gstatic.com/ipranges/cloud.json) for a region

**Comment by jpassing** [comment]
I think adding support for Secrets Manager makes sense. This could look like:

* Introduce a new environment variable `SMTP_SECRET` which you can set to a Secrets Manager path. The application's service account must have read-access to that secret.
* If `SMTP_SECRET` is set, the application loads the SMTP password from that path. If `SMTP_SECRET` is not set or empty, the application falls back to using `SMTP_PASSWORD`.

Would that sound ok?

**Comment by rojomisin** [comment]
I'm attempting to use the email MPA workflow, but receiving an error:

```The SMTP configuration is incomplete```

Is there a debug flag to see more details on the failure?

```yaml
runtime: java11
env: standard
instance_class: F2
handlers:
  - url: .*
    script: auto
env_variables:
  ACTIVATION_REQUEST_TIMEOUT: '30'
  ELEVATION_DURATION: '240'
  JUSTIFICATION_HINT: task id
  JUSTIFICATION_PATTERN: .*
  NOTIFICATION_TIMEZONE: America/Los_Angeles
  RESOURCE_SCOPE: projects/test-project1
  SMTP_ENABLE_STARTTLS: 'true'
  SMTP_HOST: smtp-relay.gmail.com
  SMTP_PASSWORD: <redacted>
  SMTP_PORT: '587'
  SMTP_SENDER_ADDRESS: jitaccess-notifications@<redacted>
  SMTP_SENDER_NAME: jit access
  SMTP_USERNAME: jitaccess-notifications@<redacted>
automatic_scaling:
  min_idle_instances: automatic
  max_idle_instances: automatic
  min_pending_latency: automatic
  max_pending_latency: automatic
service_account: jitaccess-stg@test-project1.iam.gserviceaccount.com
```


```
2023-03-09 00:58:07 jitaccess-stg[20230307t012038]  "GET /api/policy HTTP/1.1" 200
2023-03-09 00:58:07 jitaccess-stg[20230307t012038]  <JsonPayloadValue   additionalProperties: [<AdditionalProperty   key: 'logging.googleapis.com/trace'   value: <JsonValue   is_null: True>>, <AdditionalProperty   key: 'message'   value: <JsonValue   string_value: 'The SMTP configuration is incomplete'>>]>
2023-03-09 00:58:07 jitaccess-stg[20230307t012038]  <JsonPayloadValue   additionalProperties: [<AdditionalProperty   key: 'logging.googleapis.com/trace'   value: <JsonValue   is_null: True>>, <AdditionalProperty   key: 'message'   value: <JsonValue   string_value: 'Running in project test-project1 (422295094777) as jitaccess-stg@test-project1.iam.gserviceaccount.com, version 1.3.0'>>]>
2023-03-09 00:58:07 jitaccess-stg[20230307t012038]  Authenticated IAP principal
2023-03-09 00:58:09 jitaccess-stg[20230307t012038]  "GET /api/projects/scope-project1/roles HTTP/1.1" 200
2023-03-09 00:58:09 jitaccess-stg[20230307t012038]  Authenticated IAP principal
2023-03-09 00:58:13 jitaccess-stg[20230307t012038]  "GET /api/projects/scope-project1/peers?role=roles%2Fstorage.admin HTTP/1.1" 200
2023-03-09 00:58:13 jitaccess-stg[20230307t012038]  Authenticated IAP principal
2023-03-09 00:59:03 jitaccess-stg[20230307t012038]  2023-03-09 00:59:03,444 INFO  [io.quarkus] (Shutdown thread) jitaccess stopped in 0.472s
2023-03-09 01:00:17 jitaccess-stg[20230308t165605]  "POST /api/projects/scope-project1/roles/request HTTP/1.1" 400
2023-03-09 01:00:17 jitaccess-stg[20230308t165605]  <JsonPayloadValue   additionalProperties: [<AdditionalProperty   key: 'message'   value: <JsonValue   string_value: 'Running in project test-project1 (422295094777) as jitaccess-stg@test-project1.iam.gserviceaccount.com, version 1.3.0'>>, <AdditionalProperty   key: 'logging.googleapis.com/trace'   value: <JsonValue   is_null: True>>]>
2023-03-09 01:00:18 jitaccess-stg[20230308t165605]  Authenticated IAP principal
2023-03-09 01:01:04 jitaccess-stg[20230308t165605]  "POST /api/projects/scope-project1/roles/request HTTP/1.1" 403
2023-03-09 01:01:04 jitaccess-stg[20230308t165605]  Authenticated IAP principal
2023-03-09 01:01:07 jitaccess-stg[20230308t165605]  User user.test1@<redacted.com> failed to request role 'roles/storage.admin' on '//cloudresourcemanager.googleapis.com/projects/scope-project1': The notification could not be sent, caused by MailException: The mail could not be delivered, caused by MessagingException: [EOF]
2023-03-09 01:13:13 jitaccess-stg[20230307t012038]  2023-03-09 01:13:13,449 INFO  [io.quarkus] (Shutdown thread) jitaccess stopped in 0.471s
2023-03-09 01:13:23 jitaccess-stg[20230307t012038]  2023-03-09 01:13:23,351 INFO  [io.quarkus] (Shutdown thread) jitaccess stopped in 0.357s

```

**Comment by jpassing** [comment]
Was  `SMTP_SENDER_ADDRESS` not set to a valid Gmail address? Or what else did you change in `app.yaml` to fix the issue?

The _Email API authorized senders_ setting actually shouldn't have any effect: The setting only applies to the Mail API used by older runtimes. JIT Access is based on 2-gen AppEngine and doesn't use that API at all.

**Comment by rojomisin** [comment]
Our google workspace admin provisioned a `jitaccess-notifications@domain.com` user and gmail "app password" to use as the username and password vars, however I erroneously also set that user as the `SMTP_SENDER_ADDRESS`, which didn't work, which makes me think it doesn't really have an email address, or it's not a principal in a group with permission to send/relay.  Once I set the sender address to our main group "engineering@domain.com" it worked fine.

![Screenshot 2023-03-09 at 3 55 09 PM](https://user-images.githubusercontent.com/3313148/224187745-6f2e8e29-a913-404e-bfcf-9bcccd18c974.png)

**Comment by rojomisin** [comment]
i am finally getting around to using SMTP_SECRET in place of SMTP_PASSWORD, but finding an error `The SMTP configuration is incomplete`

I have `SMTP_PASSWORD` commented out, should I leave as a blank string? ""?

here's my config
```yaml
runtime: java11
env: standard
instance_class: F2
handlers:
  - url: .*
    script: auto
env_variables:
  ACTIVATION_REQUEST_TIMEOUT: '30'
  ELEVATION_DURATION: '60'
  JUSTIFICATION_HINT: 'Gitlab MR or Linear task ID -- NOTE: Only 1 approval is required'
  JUSTIFICATION_PATTERN: .*
  NOTIFICATION_TIMEZONE: America/Los_Angeles
  RESOURCE_SCOPE: projects/production
  SMTP_ENABLE_STARTTLS: 'true'
  SMTP_HOST: smtp.gmail.com
  SMTP_PORT: '587'
  SMTP_SECRET: >-
    projects/<myprojectnumber>/secrets/jitaccess-notifications-smtp-password/versions/latest
  SMTP_SENDER_ADDRESS: engineering@domain.com
  SMTP_SENDER_NAME: JIT Auth (production)
  SMTP_USERNAME: jitaccess-notifications@domaincom

```

**Comment by rojomisin** [comment]
Confirmed that there is now not a combination of using SMTP_SECRET and/or SMTP_PASSWORD that works for MPA email.  It will always log a warning `The SMTP configuration is incomplete` and the app itself then fails to find any roles., but perhaps that's a symptom of the email configuration not passing.
<img width="804" alt="Screenshot 2023-03-30 at 4 30 08 PM" src="https://user-images.githubusercontent.com/3313148/228987346-c79ab4bb-1b29-441b-8988-80f40440c80b.png">

The non-email self-approval workflow in our other environments still works fine w/ master build `39196539bbaa6919be7be315692149e052ac01bd`

**Comment by jpassing** [comment]
The error `The SMTP configuration is incomplete` is logged when one of the following settings is missing or invalid:

* SMTP_HOST
* SMTP_PORT
* SMTP_SENDER_NAME
* SMTP_SENDER_ADDRESS

The MPA feature is then disabled, and the app doesn't even attempt to read the SMTP password or secret. Other functionality (in particular, listing eligible roles) shouldn't depend on the email configuration in any way.

I noticed that your configuration above is missing the `service_account:` line. Could you double-check that this line is present in your deployment config? Otherwise, that might explain the strange behavior.

The rest of your configuration above looks ok to me. I though the `>-` might be causing trouble, but in my tests that worked fine too. Maybe it's still worth double-checking that there aren't any indentation differences between what you pasted above and what you were using.

> I have SMTP_PASSWORD commented out, should I leave as a blank string? ""?

Both approaches should be ok.

**Comment by rojomisin** [comment]
here's a screenshot of my `app.yaml`, all fields are there, my editor is not making any yaml mistakes, but I still cannot get it to work 🤔 

my SMTP_PASSWORD is a valid passcode (16 character). 

<img width="847" alt="Screenshot 2023-04-03 at 3 20 29 PM" src="https://user-images.githubusercontent.com/3313148/229639994-8704f167-0245-47de-9608-1cf73476697b.png">

**Comment by rojomisin** [comment]
confirmed google group read permissions for the project was missing, so a red herring error in the logs, sorry about that

**Comment by jpassing** [comment]
> confirmed google group read permissions for the project was missing

Does everything work again as expected then? Or are you still seeing the `The SMTP configuration is incomplete` error in the logs?

**Comment by rojomisin** [comment]
no worries, ty for confirming

**Comment by jpassing** [comment]
This feature is now available in [release 1.4](https://github.com/GoogleCloudPlatform/jit-access/releases/tag/1.4.0), and I updated the [deployment instructions](https://github.com/GoogleCloudPlatform/jit-access/wiki/Configure-Multi-Party-Approval) to use `SMTP_SECRET` instead of `SMTP_PASSWORD`.
